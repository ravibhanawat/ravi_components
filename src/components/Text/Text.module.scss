@import "../../styles/themes";
@import "../../styles/typography";
@import "../../styles/states";

.text {
  a {
    text-decoration: none;
    color: var(--link-color);
    @include focus-style();

    &:hover {
      text-decoration: underline;
    }
  }

  &.primary {
    color: var(--primary-text-color);
  }

  &.secondary {
    color: var(--secondary-text-color);
  }

  &.onPrimary {
    color: var(--text-color-on-primary);
  }

  &.onInverted {
    color: var(--text-color-on-inverted);
  }

  &.start {
    text-align: start;
  }

  &.center {
    text-align: center;
  }

  &.end {
    text-align: end;
  }

  &.singleLineEllipsis {
    @include ellipsis();
  }

  &.multiLineEllipsis {
    @include line-clamp(var(--text-clamp-lines));
  }
}
