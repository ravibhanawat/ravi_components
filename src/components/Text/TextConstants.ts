export type TextSize = "medium" | "small";

export type TextWeight = "bold" | "normal";

export type TextColor = "primary" | "secondary" | "onPrimary" | "onInverted";

export type TextAlign = "start" | "center" | "end";
