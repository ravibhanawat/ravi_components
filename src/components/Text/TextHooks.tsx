import { MutableRefObject } from "react";
import { TextSize, TextWeight } from "./TextConstants";
import { ElementContent } from "../../types";
import useIsOverflowing from "../../hooks/useIsOverflowing/useIsOverflowing";
import { TooltipProps } from "../Tooltip/Tooltip";
import styles from "./Text.module.scss";

export function useGlobalTextClass(size: TextSize, weight: TextWeight) {
  return `vibe-text-${size}-${weight}`;
}

export function useEllipsisClass(ref: (node: HTMLElement) => void, ellipsis: boolean, maxLines: number) {
  let ellipsisClass;
  const overrideRef = (node: HTMLElement) => {
    node?.style.setProperty("--text-clamp-lines", maxLines.toString());
    ref(node);
  };

  // If component contains ellipsis return the fit ellipsis class
  if (ellipsis) {
    ellipsisClass = maxLines > 1 ? styles.multiLineEllipsis : styles.singleLineEllipsis;
  }

  return { ref: overrideRef, class: ellipsisClass };
}

export function useTooltipProps(
  ref: MutableRefObject<any>,
  withoutTooltip: boolean,
  ellipsis: boolean,
  tooltipProps: TooltipProps,
  children: ElementContent
) {
  const isOverflowing = useIsOverflowing({ ref: ellipsis ? ref : null, ignoreHeightOverflow: true });
  const isTooltipRendered = !withoutTooltip && ellipsis && isOverflowing && typeof children === "string";
  return isTooltipRendered ? { ...tooltipProps, content: children } : {};
}
