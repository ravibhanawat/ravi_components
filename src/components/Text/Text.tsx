import React, { forwardRef, useRef } from "react";
import cx from "classnames";
import { useMergeRefs } from "../../hooks";
import VibeComponentProps from "../../types/VibeComponentProps";
import VibeComponent from "../../types/VibeComponent";
import { getTestId } from "../../tests/test-ids-utils";
import { ComponentDefaultTestId } from "../../tests/constants";
import { ElementContent } from "../../types";
import { TextSize, TextWeight, TextColor, TextAlign } from "./TextConstants";
import { useEllipsisClass, useGlobalTextClass, useTooltipProps } from "./TextHooks";
import Tooltip, { TooltipProps } from "../Tooltip/Tooltip";
import styles from "./Text.module.scss";

export interface TextProps extends VibeComponentProps {
  /**
   * The element tag of the text component
   */
  element?: string;
  /**
   * The textual content
   */
  children: ElementContent;
  size?: TextSize;
  weight?: TextWeight;
  color?: TextColor;
  align?: TextAlign;
  /**
   * When the text is too long, cut the end of the text and display instead of it three dots (...)
   */
  ellipsis?: boolean;
  /**
   * Use this prop combined with the boolean ellipsis prop for truncate the text and add an ellipsis after a certain number of lines
   */
  maxLines?: number;
  /**
   * All props are passed to the tooltip displayed when hovering over the text. By default, the tooltip will display when text contains an ellipsis and will show the full text
   */
  tooltipProps?: TooltipProps;
  /**
   * Hide tooltip when hovering the text, by default the tooltip swill display when text contains an ellipsis
   */
  withoutTooltip?: boolean;
}

const Text: VibeComponent<TextProps, HTMLElement> = forwardRef(
  (
    {
      className,
      id,
      children,
      tooltipProps,
      "data-testid": dataTestId = getTestId(ComponentDefaultTestId.TEXT, id),
      element = "span",
      size = "medium",
      weight = "normal",
      color = "primary",
      align = "start",
      ellipsis = true,
      maxLines = 1,
      withoutTooltip = false
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    const textGlobalClass = useGlobalTextClass(size, weight);
    const { ref: overrideRef, class: ellipsisClass } = useEllipsisClass(mergedRef, ellipsis, maxLines);
    const overrideTooltipProps = useTooltipProps(
      componentRef,
      withoutTooltip,
      ellipsis,
      tooltipProps,
      children
    ) as TooltipProps;

    return (
      <Tooltip {...overrideTooltipProps}>
        {React.createElement(
          element,
          {
            id,
            "data-testid": dataTestId,
            className: cx(textGlobalClass, styles.text, styles[color], styles[align], ellipsisClass, className),
            ref: overrideRef
          },
          children
        )}
      </Tooltip>
    );
  }
);

export default Text;
