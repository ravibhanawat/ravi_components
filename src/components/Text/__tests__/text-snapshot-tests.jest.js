import React from "react";
import renderer from "react-test-renderer";
import Text from "../Text";

describe("Text renders correctly", () => {
  it("with empty props", () => {
    const tree = renderer.create(<Text />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with size", () => {
    const tree = renderer.create(<Text size="small">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with weight", () => {
    const tree = renderer.create(<Text weight="bold">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with color", () => {
    const tree = renderer.create(<Text color="secondary">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with element", () => {
    const tree = renderer.create(<Text element="p">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with align", () => {
    const tree = renderer.create(<Text align="center">text</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with element as children", () => {
    const tree = renderer
      .create(
        <Text>
          <div>text</div>
        </Text>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
