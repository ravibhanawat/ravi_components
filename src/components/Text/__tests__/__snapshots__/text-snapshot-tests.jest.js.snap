// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly with align 1`] = `
<span
  className="vibe-text-medium-normal text primary center singleLineEllipsis"
  data-testid="text"
>
  text
</span>
`;

exports[`Text renders correctly with color 1`] = `
<span
  className="vibe-text-medium-normal text secondary start singleLineEllipsis"
  data-testid="text"
>
  text
</span>
`;

exports[`Text renders correctly with element 1`] = `
<p
  className="vibe-text-medium-normal text primary start singleLineEllipsis"
  data-testid="text"
>
  text
</p>
`;

exports[`Text renders correctly with element as children 1`] = `
<span
  className="vibe-text-medium-normal text primary start singleLineEllipsis"
  data-testid="text"
>
  <div>
    text
  </div>
</span>
`;

exports[`Text renders correctly with empty props 1`] = `
<span
  className="vibe-text-medium-normal text primary start singleLineEllipsis"
  data-testid="text"
/>
`;

exports[`Text renders correctly with size 1`] = `
<span
  className="vibe-text-small-normal text primary start singleLineEllipsis"
  data-testid="text"
>
  text
</span>
`;

exports[`Text renders correctly with weight 1`] = `
<span
  className="vibe-text-medium-bold text primary start singleLineEllipsis"
  data-testid="text"
>
  text
</span>
`;
